[MM-14823] Render emojis in Interactive Message Buttons

kosgrz opened the pull request:
#### Summary
This PR makes it possible to render emojis in Interactive Message Buttons.

#### Ticket Link
[Jira ticket](https://mattermost.atlassian.net/browse/MM-14823) | Fixes [GitHub issue #10290](https://github.com/mattermost/mattermost-server/issues/10290)

#### Checklist
N/A

#### Device 
This PR was tested on: iPhone X with iOS 12.1, Nexus 5X with Android 9.0

#### Screenshots
![iphone](https://user-images.githubusercontent.com/45372453/54091566-b9e2c700-4381-11e9-9026-b8da7a91d7d0.png)
![android](https://user-images.githubusercontent.com/45372453/54091568-bcddb780-4381-11e9-9ea3-bb541b84e4f6.png)

sudheerDev commented:
webapp PR is merged we can resolve the ticket when this is merged

enahum commented:
This will support more than emojis is supporting almost everything on markdown, is that what is expected here?

enahum commented:
Please link the webapp PR

hanzei commented:
@enahum https://github.com/mattermost/mattermost-webapp/pull/2476

hmhealey commented on app/components/message_attachments/action_button/action_button.js:
```diff
@@ -34,7 +34,19 @@ export default class ActionButton extends PureComponent {
             >
-                <Text style={style.text}>{name}</Text>
+                <Markdown
+                    isEdited={false}
```
Not related to this PR, but we really need to rethink how the `Markdown` component is structured since this is way more complicated than it should be.

enahum commented:
@hmhealey I’m not really sure.
@esethna thoughts?

hmhealey commented:
The alignment might be a tough one to deal with since it's likely the same as the alignment issues with emojis in regular post text. Perhaps there's something we can do with modifying the `baseTextStyle` to fix that

kosgrz commented:
Thank you all for the review. I was extremely busy last week, but now I will have more free time and I think I update the PR today or tomorrow with a new component to render markdown in Interactive Message Buttons.

@esethna sorry that I didn't notice the alignment issue when I was creating the PR :/

> The alignment might be a tough one to deal with since it's likely the same as the alignment issues with emojis in regular post text. Perhaps there's something we can do with modifying the baseTextStyle to fix that

@hmhealey I think I was dealing with a very similar issue when was working on another PR in the past. Indeed modifying the baseTextStyle might fix it :)

jasonblais commented:
@kosgrz Does the webapp PR also support other Markdown in the buttons, in addition to just plain text and emoji? https://github.com/mattermost/mattermost-webapp/pull/2476

kosgrz commented:
@jasonblais, fortunately, the webapp PR support only emojis. Everything else is rendered as a plain text. Some examples:
![image](https://user-images.githubusercontent.com/45372453/54860703-32e90380-4d1e-11e9-9afb-5099e204ed55.png)
![image](https://user-images.githubusercontent.com/45372453/54860720-6592fc00-4d1e-11e9-91d2-accfd58a6d2c.png)
![image](https://user-images.githubusercontent.com/45372453/54860731-8bb89c00-4d1e-11e9-88ef-85ff670c1d56.png)

kosgrz commented:
I updated the PR. Now it should only render emojis. I also fixed the alignment issue with iPhone devices. Some screenshots:
![Screenshot_1553395939](https://user-images.githubusercontent.com/45372453/54874291-4f517280-4de8-11e9-896c-a127567cc9b3.png)
![iphone](https://user-images.githubusercontent.com/45372453/54874276-17e2c600-4de8-11e9-8d73-712f4b20cda1.png)
![android](https://user-images.githubusercontent.com/45372453/54874278-1ca77a00-4de8-11e9-8e20-fdcc74cb9c41.png)

kosgrz commented on app/components/message_attachments/action_button/index.js:
```diff
@@ -6,12 +6,18 @@ import {connect} from 'react-redux';
 
-function mapStateToProps(state) {
+function mapStateToProps(state, ownProps) {
+    const memoizeHasAnyEmojis = memoizeResult((message, customEmojis) => hasAnyEmojis(message, customEmojis));
```
Thanks for the explanation!

kosgrz commented:
@hmhealey I updated the PR. Thanks for your guidance, it was really helpful! 
I didn't delete your comments, because I think it can be useful in the future.
I also changed regex for the ":`(" emoticon in emoji_utils.js, because the old didn't work in my case. Now it is the same as in the webapp.

Some screens:
![iphone](https://user-images.githubusercontent.com/45372453/55150585-b787b780-514c-11e9-9ceb-3d168e4f0141.png)
![android](https://user-images.githubusercontent.com/45372453/55150586-b787b780-514c-11e9-8883-b205f2e04d5e.png)

hmhealey commented on app/components/message_attachments/action_button/action_button_text.js:
```diff
@@ -0,0 +1,89 @@
+// before the next emoji by looking for any character that could start an emoji (:, ;, x, or <)
+const reMain = /^[\s\S]+?(?=[:;x<]|$)/i;
+
+const containerStyles = {
```
Could you change this to follow our other components where we have
```
import {StyleSheet} from 'react-native';

export default function ActionButtonText() {
    return <View style={styles.container}/>;
}

const styles = StyleSheet.create({
    container: {
        alignItems: 'flex-start',
        flexDirection: 'row',
        flexWrap: 'wrap',
    },
});
```
Using a StyleSheet is supposed to slightly help performance, and changing this would also make the code a bit more consistent.

kosgrz replied:
Done.

hmhealey commented on app/components/message_attachments/action_button/action_button_text.js:
```diff
@@ -0,0 +1,89 @@
+    flexWrap: 'wrap',
+};
+
+export function ActionButtonText({message, style}) {
```
Can you change this so that it's the default export to be consistent with other components as well?

kosgrz replied:
Done.

jasonblais commented:
FYI - I created a separate ticket for mobile, so we can track the mobile and webapp PRs separately: https://mattermost.atlassian.net/browse/MM-14823

Reason is that I resolved the webapp ticket so our QA team can test and close it for the 5.10 release.

hanzei commented:
I can't merge this PR. Would someone else please take care of this?

hmhealey commented:
Yep, I'll do that now

hmhealey commented:
Thanks for working through this, @kosgrz, even though it required a rewrite or two as issues came up :smile:
